File: src/ukf.cpp
#include "kalman_positioning/ukf.hpp"
#include <iostream>
#include <map>

/**
 * STUDENT ASSIGNMENT: Unscented Kalman Filter Implementation
 * 
 * This file contains placeholder implementations for the UKF class methods.
 * Students should implement each method according to the UKF algorithm.
 * 
 * Reference: Wan, E. A., & Van Der Merwe, R. (2000). 
 * "The Unscented Kalman Filter for Nonlinear Estimation"
 */

// ============================================================================
// CONSTRUCTOR
// ============================================================================

/**
 * @brief Initialize the Unscented Kalman Filter
 * 
 * STUDENT TODO:
 * 1. Initialize filter parameters (alpha, beta, kappa, lambda)
 * 2. Initialize state vector x_ with zeros
 * 3. Initialize state covariance matrix P_ 
 * 4. Set process noise covariance Q_
 * 5. Set measurement noise covariance R_
 * 6. Calculate sigma point weights for mean and covariance
 */
UKF::UKF(double process_noise_xy, double process_noise_theta,
         double measurement_noise_xy, int num_landmarks)
    : nx_(5), nz_(2) {
    
    // Task A1: UKF Constructor and Initialization
    // ========================================================================
    
    // 1. Initialize state vector x = [0, 0, 0, 0, 0]
    x_ = Eigen::VectorXd::Zero(nx_);
    
    // 2. Initialize P as identity matrix (Task A1)
    P_ = Eigen::MatrixXd::Identity(nx_, nx_);
    
    // 3. Set Q = diag(process_noise_xy, process_noise_xy, process_noise_theta, 0, 0) (Task A1)
    Q_ = Eigen::MatrixXd::Zero(nx_, nx_);
    Q_(0, 0) = process_noise_xy;      // x
    Q_(1, 1) = process_noise_xy;      // y
    Q_(2, 2) = process_noise_theta;   // theta
    Q_(3, 3) = 0.0;                   // vx (as specified)
    Q_(4, 4) = 0.0;                   // vy (as specified)
    
    // 4. Set measurement noise covariance R
    R_ = Eigen::MatrixXd::Identity(nz_, nz_) * measurement_noise_xy;
    
    // 5. Calculate UKF parameters: λ, γ, and weights (Task A1)
    // Using parameters that ensure positive weights for stability
    double alpha = 1.0;   // Spread parameter
    double beta = 0.0;    // Using 0 instead of 2 to avoid huge W0_c
    double kappa = 2.0;   // Secondary scaling (3-n would be -2, use 2 instead)
    
    lambda_ = alpha * alpha * (nx_ + kappa) - nx_;  // = 1*(5+2) - 5 = 2
    gamma_ = std::sqrt(nx_ + lambda_);               // = sqrt(7) = 2.646
    
    // 6. Calculate weights W^m_i, W^c_i (Task A1)
    Wm_.resize(2 * nx_ + 1);
    Wc_.resize(2 * nx_ + 1);
    
    Wm_[0] = lambda_ / (nx_ + lambda_);                                    // = 2/7 = 0.286
    Wc_[0] = lambda_ / (nx_ + lambda_) + (1.0 - alpha * alpha + beta);   // = 2/7 + 0 = 0.286
    
    for (int i = 1; i < 2 * nx_ + 1; i++) {
        Wm_[i] = 1.0 / (2.0 * (nx_ + lambda_));  // = 1/14 = 0.071
        Wc_[i] = 1.0 / (2.0 * (nx_ + lambda_));  // = 1/14 = 0.071
    }
    
    std::cout << "UKF Initialized: lambda=" << lambda_ << ", gamma=" << gamma_ << std::endl;
}

// ============================================================================
// SIGMA POINT GENERATION
// ============================================================================

/**
 * @brief Generate sigma points from mean and covariance
 * 
 * STUDENT TODO:
 * 1. Start with the mean as the first sigma point
 * 2. Compute Cholesky decomposition of covariance
 * 3. Generate 2*n symmetric sigma points around the mean
 */
std::vector<Eigen::VectorXd> UKF::generateSigmaPoints(const Eigen::VectorXd& mean,
                                                       const Eigen::MatrixXd& cov) {
    // Task A2: Sigma Point Generation (NO CHOLESKY - using eigenvalue decomposition)
    // ========================================================================
    
    std::vector<Eigen::VectorXd> sigma_points;
    int n = mean.size();
    
    // 1. First sigma point is the mean
    sigma_points.push_back(mean);
    
    // 2. Compute matrix square root using eigenvalue decomposition (Task A2)
    // Instead of Cholesky: P = V * D * V^T, so sqrt(P) = V * sqrt(D)
    
    Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> solver(cov);
    Eigen::VectorXd eigenvalues = solver.eigenvalues();
    Eigen::MatrixXd eigenvectors = solver.eigenvectors();
    
    // Ensure positive eigenvalues for numerical stability
    for (int i = 0; i < eigenvalues.size(); i++) {
        if (eigenvalues(i) < 1e-10) eigenvalues(i) = 1e-10;
    }
    
    // Compute sqrt(P) = V * sqrt(D), then scale by gamma
    // This gives us γ*sqrt(P) as required for sigma points
    Eigen::MatrixXd sqrt_matrix = eigenvectors * eigenvalues.cwiseSqrt().asDiagonal();
    
    // 3. Generate 2*n sigma points: Xi = x ± γ*Li (Task A2)
    for (int i = 0; i < n; i++) {
        sigma_points.push_back(mean + gamma_ * sqrt_matrix.col(i));  // Positive direction
    }
    for (int i = 0; i < n; i++) {
        sigma_points.push_back(mean - gamma_ * sqrt_matrix.col(i));  // Negative direction
    }
    
    return sigma_points;
}

// ============================================================================
// PROCESS MODEL
// ============================================================================

/**
 * @brief Apply motion model to a state vector
 * 
 * STUDENT TODO:
 * 1. Updates position: x' = x + dx, y' = y + dy
 * 2. Updates orientation: theta' = theta + dtheta (normalized)
 * 3. Updates velocities: vx' = dx/dt, vy' = dy/dt
 */
Eigen::VectorXd UKF::processModel(const Eigen::VectorXd& state, double dt,
                                  double dx, double dy, double dtheta) {
    // Task A3: Process Model
    // ========================================================================
    
    Eigen::VectorXd new_state(5);
    
    // Update position: x' = x + dx, y' = y + dy
    new_state(0) = state(0) + dx;
    new_state(1) = state(1) + dy;
    
    // Update orientation: theta' = theta + dtheta (normalized)
    new_state(2) = normalizeAngle(state(2) + dtheta);
    
    // Update velocities: vx = dx/dt, vy = dy/dt
    if (dt > 1e-6) {
        new_state(3) = dx / dt;
        new_state(4) = dy / dt;
    } else {
        new_state(3) = 0.0;
        new_state(4) = 0.0;
    }
    
    return new_state;
}

// ============================================================================
// MEASUREMENT MODEL
// ============================================================================

/**
 * @brief Predict measurement given current state and landmark
 * 
 * STUDENT TODO:
 * 1. Calculate relative position: landmark - robot position
 * 2. Transform to robot frame using robot orientation
 * 3. Return relative position in robot frame
 */
Eigen::Vector2d UKF::measurementModel(const Eigen::VectorXd& state, int landmark_id) {
    // Task A4: Measurement Model
    // ========================================================================
    // The observations from fake_robot are ALREADY in robot frame (relative x, y)
    // So we need to predict what the robot WOULD observe given its state
    
    if (landmarks_.find(landmark_id) == landmarks_.end()) {
        return Eigen::Vector2d::Zero();
    }
    
    // Get landmark position in world frame
    auto landmark = landmarks_[landmark_id];
    double lx = landmark.first;
    double ly = landmark.second;
    
    // Get robot state in world frame
    double rx = state(0);
    double ry = state(1);
    double theta = state(2);
    
    // Calculate relative position in world frame
    double dx_world = lx - rx;
    double dy_world = ly - ry;
    
    // Transform to robot frame (what the robot's sensor would see)
    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);
    
    Eigen::Vector2d predicted_observation;
    // Rotation from world to robot frame
    predicted_observation(0) = cos_theta * dx_world + sin_theta * dy_world;   // forward
    predicted_observation(1) = -sin_theta * dx_world + cos_theta * dy_world;  // sideways
    
    return predicted_observation;
}

// ============================================================================
// ANGLE NORMALIZATION
// ============================================================================

double UKF::normalizeAngle(double angle) {
    while (angle > M_PI) angle -= 2.0 * M_PI;
    while (angle < -M_PI) angle += 2.0 * M_PI;
    return angle;
}

// ============================================================================
// PREDICTION STEP
// ============================================================================

/**
 * @brief Kalman Filter Prediction Step (Time Update)
 * 
 * STUDENT TODO:
 * 1. Generate sigma points from current state and covariance
 * 2. Propagate each sigma point through motion model
 * 3. Calculate mean and covariance of predicted sigma points
 * 4. Add process noise
 * 5. Update state and covariance estimates
 */
void UKF::predict(double dt, double dx, double dy, double dtheta) {
    // Task A5: Predict Step
    // ========================================================================
    
    // 1. Generate sigma points from current state and covariance
    std::vector<Eigen::VectorXd> sigma_points = generateSigmaPoints(x_, P_);
    
    // 2. Propagate each sigma point through process model
    std::vector<Eigen::VectorXd> sigma_points_pred;
    for (const auto& sp : sigma_points) {
        sigma_points_pred.push_back(processModel(sp, dt, dx, dy, dtheta));
    }
    
    // 3. Calculate predicted mean
    Eigen::VectorXd x_pred = Eigen::VectorXd::Zero(nx_);
    for (size_t i = 0; i < sigma_points_pred.size(); i++) {
        x_pred += Wm_[i] * sigma_points_pred[i];
    }
    x_pred(2) = normalizeAngle(x_pred(2));  // Normalize angle
    
    // 4. Calculate predicted covariance
    Eigen::MatrixXd P_pred = Eigen::MatrixXd::Zero(nx_, nx_);
    for (size_t i = 0; i < sigma_points_pred.size(); i++) {
        Eigen::VectorXd diff = sigma_points_pred[i] - x_pred;
        diff(2) = normalizeAngle(diff(2));  // Normalize angle difference
        P_pred += Wc_[i] * (diff * diff.transpose());
    }
    
    // 5. Add process noise
    P_pred += Q_;
    
    // Update state and covariance
    x_ = x_pred;
    P_ = P_pred;
}

// ============================================================================
// UPDATE STEP
// ============================================================================

/**
 * @brief Kalman Filter Update Step (Measurement Update)
 * 
 * STUDENT TODO:
 * 1. Generate sigma points
 * 2. Transform through measurement model
 * 3. Calculate predicted measurement mean
 * 4. Calculate measurement and cross-covariance
 * 5. Compute Kalman gain
 * 6. Update state with innovation
 * 7. Update covariance
 */
void UKF::update(const std::vector<std::tuple<int, double, double, double>>& landmark_observations) {
    if (landmark_observations.empty()) {
        return;
    }
    
    // Task A6: Standard UKF Update - One landmark at a time
    // ========================================================================
    
    std::cout << "\n=== UKF UPDATE ===" << std::endl;
    std::cout << "Current state: x=" << x_(0) << ", y=" << x_(1) << ", theta=" << x_(2) << std::endl;
    
    for (const auto& obs : landmark_observations) {
        int landmark_id = std::get<0>(obs);
        double obs_x = std::get<1>(obs);
        double obs_y = std::get<2>(obs);
        
        if (!hasLandmark(landmark_id)) {
            continue;
        }
        
        auto lm = landmarks_[landmark_id];
        std::cout << "  LM" << landmark_id << " world=(" << lm.first << "," << lm.second << ")" << std::endl;
        
        // Debug: Show what we SHOULD observe vs what we DID observe
        Eigen::Vector2d predicted = measurementModel(x_, landmark_id);
        std::cout << "    Predicted obs (robot frame): (" << predicted(0) << "," << predicted(1) << ")" << std::endl;
        std::cout << "    Actual obs    (robot frame): (" << obs_x << "," << obs_y << ")" << std::endl;
        std::cout << "    Innovation: (" << (obs_x - predicted(0)) << "," << (obs_y - predicted(1)) << ")" << std::endl;
        
        // 1. Generate sigma points
        std::vector<Eigen::VectorXd> sigma_points = generateSigmaPoints(x_, P_);
        
        // 2. Transform through measurement model
        std::vector<Eigen::Vector2d> Z_sigma;
        for (const auto& sp : sigma_points) {
            Z_sigma.push_back(measurementModel(sp, landmark_id));
        }
        
        // 3. Calculate predicted measurement mean
        Eigen::Vector2d z_mean = Eigen::Vector2d::Zero();
        for (size_t i = 0; i < Z_sigma.size(); i++) {
            z_mean += Wm_[i] * Z_sigma[i];
        }
        
        // 4. Calculate innovation covariance and cross-covariance
        Eigen::MatrixXd P_zz = Eigen::MatrixXd::Zero(nz_, nz_);
        Eigen::MatrixXd P_xz = Eigen::MatrixXd::Zero(nx_, nz_);
        
        for (size_t i = 0; i < sigma_points.size(); i++) {
            Eigen::Vector2d z_diff = Z_sigma[i] - z_mean;
            P_zz += Wc_[i] * (z_diff * z_diff.transpose());
            
            Eigen::VectorXd x_diff = sigma_points[i] - x_;
            x_diff(2) = normalizeAngle(x_diff(2));
            P_xz += Wc_[i] * (x_diff * z_diff.transpose());
        }
        
        // Add measurement noise
        P_zz += R_;
        
        // 5. Check for numerical issues
        if (P_zz.determinant() < 1e-10) {
            continue;  // Skip this update if covariance is singular
        }
        
        // 6. Compute Kalman gain
        Eigen::MatrixXd K = P_xz * P_zz.inverse();
        
        // 7. Update state with innovation
        Eigen::Vector2d z_obs(obs_x, obs_y);
        Eigen::Vector2d innovation = z_obs - z_mean;
        
        // Limit innovation to prevent huge jumps (optional safety check)
        double innovation_norm = innovation.norm();
        if (innovation_norm > 5.0) {  // Max 5m correction per landmark
            innovation = innovation / innovation_norm * 5.0;
        }
        
        Eigen::VectorXd correction = K * innovation;
        std::cout << "    Kalman gain correction: dx=" << correction(0) << ", dy=" << correction(1) << ", dtheta=" << correction(2) << std::endl;
        
        x_ = x_ + correction;
        x_(2) = normalizeAngle(x_(2));
        
        std::cout << "    New state: x=" << x_(0) << ", y=" << x_(1) << ", theta=" << x_(2) << std::endl;
        
        // 8. Update covariance
        P_ = P_ - K * P_zz * K.transpose();
        P_ = (P_ + P_.transpose()) / 2.0;  // Symmetrize
        
        // Add regularization
        P_ += Eigen::MatrixXd::Identity(nx_, nx_) * 1e-6;
    }
    
    std::cout << "=== END UPDATE ===" << std::endl;
}

// ============================================================================
// LANDMARK MANAGEMENT
// ============================================================================

void UKF::setLandmarks(const std::map<int, std::pair<double, double>>& landmarks) {
    landmarks_ = landmarks;
}

bool UKF::hasLandmark(int id) const {
    return landmarks_.find(id) != landmarks_.end();
}

File: python/kalman_positioning/ukf.py
"""
STUDENT ASSIGNMENT: Unscented Kalman Filter Implementation

This file contains the complete UKF implementation for robot localization.
Students should implement each method according to the UKF algorithm.

Reference: Wan, E. A., & Van Der Merwe, R. (2000). 
"The Unscented Kalman Filter for Nonlinear Estimation"
"""

import numpy as np
from typing import Dict, List, Tuple


class UKF:
    """
    Unscented Kalman Filter for robot localization with landmarks
    State: [x, y, theta, vx, vy]
    """
    
    # UKF parameters (constants)
    # Note: Alpha=0.1, Beta=2.0, Kappa=0 are standard values
    # These result in lambda=-4.95, gamma=0.224 for n=5 dimensions
    ALPHA = 0.1   # Spread of sigma points (0.001 to 1, larger = more spread)
    BETA = 2.0    # Distribution info (Gaussian = 2.0)
    KAPPA = 0.0   # Secondary scaling parameter
    
    def __init__(self, process_noise_xy: float, process_noise_theta: float,
                 measurement_noise_xy: float, num_landmarks: int = 0):
        """
        Initialize the Unscented Kalman Filter
        
        Task A1: UKF Constructor and Initialization (10 Points)
        
        STUDENT TODO:
        1. Initialize filter parameters (alpha, beta, kappa, lambda)
        2. Initialize state vector x_ with zeros
        3. Initialize state covariance matrix P_ 
        4. Set process noise covariance Q_
        5. Set measurement noise covariance R_
        6. Calculate sigma point weights for mean and covariance
        
        Args:
            process_noise_xy: Process noise for x, y positions
            process_noise_theta: Process noise for orientation
            measurement_noise_xy: Measurement noise for landmark observations
            num_landmarks: Number of landmarks (optional, for reference)
        """
        # STUDENT IMPLEMENTATION - Task A1
        # ====================================================================
        
        # State dimension (x, y, theta, vx, vy)
        self.nx = 5
        
        # Measurement dimension (x, y relative to robot)
        self.nz = 2
        
        # Initialize state vector: [x, y, theta, vx, vy]
        self.x = np.zeros(self.nx)
        
        # Initialize state covariance matrix with high initial uncertainty
        # Since we start at [0,0,0,0,0] but don't know true position
        self.P = np.diag([100.0, 100.0, 10.0, 1.0, 1.0])  # High position uncertainty initially
        
        # Process noise covariance Q
        # [x, y, theta, vx, vy]
        self.Q = np.diag([
            process_noise_xy,      # x position noise
            process_noise_xy,      # y position noise
            process_noise_theta,   # theta orientation noise
            0.0,                   # vx velocity (determined by motion)
            0.0                    # vy velocity (determined by motion)
        ])
        
        # Measurement noise covariance R (2x2 for x, y observations)
        self.R = np.diag([
            measurement_noise_xy,  # x observation noise
            measurement_noise_xy   # y observation noise
        ])
        
        # Calculate UKF parameters
        # lambda = alpha^2 * (n + kappa) - n
        self.lambda_ = self.ALPHA**2 * (self.nx + self.KAPPA) - self.nx
        
        # gamma = sqrt(n + lambda)
        self.gamma = np.sqrt(self.nx + self.lambda_)
        
        # Calculate weights for mean (Wm) and covariance (Wc)
        self.Wm = np.zeros(2 * self.nx + 1)
        self.Wc = np.zeros(2 * self.nx + 1)
        
        # Weight for mean of first sigma point
        self.Wm[0] = self.lambda_ / (self.nx + self.lambda_)
        
        # Weight for covariance of first sigma point
        self.Wc[0] = self.lambda_ / (self.nx + self.lambda_) + (1 - self.ALPHA**2 + self.BETA)
        
        # Weights for remaining sigma points (same for mean and covariance)
        for i in range(1, 2 * self.nx + 1):
            self.Wm[i] = 1.0 / (2.0 * (self.nx + self.lambda_))
            self.Wc[i] = 1.0 / (2.0 * (self.nx + self.lambda_))
        
        # Landmark storage
        self.landmarks: Dict[int, Tuple[float, float]] = {}
        
        print(f"UKF Initialized:")
        print(f"  State dimension: {self.nx}")
        print(f"  Lambda: {self.lambda_:.6f}")
        print(f"  Gamma: {self.gamma:.6f}")
        print(f"  Weights Wm: {self.Wm}")
        print(f"  Weights Wc: {self.Wc}")
    
    def generate_sigma_points(self, mean: np.ndarray, cov: np.ndarray) -> List[np.ndarray]:
        """
        Generate sigma points from mean and covariance
        
        Task A2: Sigma Point Generation (8 Points)
        
        STUDENT TODO:
        1. Start with the mean as the first sigma point
        2. Compute Cholesky decomposition of covariance
        3. Generate 2*n symmetric sigma points around the mean
        
        Args:
            mean: State mean vector (nx,)
            cov: State covariance matrix (nx, nx)
            
        Returns:
            List of 2*n+1 sigma points, each of dimension (nx,)
        """
        # STUDENT IMPLEMENTATION - Task A2
        # ====================================================================
        
        n = len(mean)
        sigma_points = []
        
        # First sigma point is the mean itself
        sigma_points.append(mean.copy())
        
        # Compute matrix square root using eigenvalue decomposition
        # This is more robust than Cholesky and doesn't require positive definiteness
        # P = U * D * U^T, so sqrt(P) = U * sqrt(D) * U^T
        
        # Scale covariance matrix
        scaled_cov = (n + self.lambda_) * cov
        
        # Eigenvalue decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(scaled_cov)
        
        # Ensure positive eigenvalues (add small epsilon if needed)
        eigenvalues = np.maximum(eigenvalues, 1e-10)
        
        # Compute matrix square root: sqrt(P) = U * sqrt(D)
        sqrt_matrix = eigenvectors @ np.diag(np.sqrt(eigenvalues))
        
        # Generate 2*n sigma points using columns of sqrt_matrix
        for i in range(n):
            # Positive direction: mean + column i of sqrt_matrix
            sigma_points.append(mean + sqrt_matrix[:, i])
            
        for i in range(n):
            # Negative direction: mean - column i of sqrt_matrix
            sigma_points.append(mean - sqrt_matrix[:, i])
        
        return sigma_points
    
    def process_model(self, state: np.ndarray, dt: float,
                      dx: float, dy: float, dtheta: float) -> np.ndarray:
        """
        Apply motion model to a state vector
        
        Task A3: Process Model (7 Points)
        
        STUDENT TODO:
        1. Updates position: x' = x + dx, y' = y + dy
        2. Updates orientation: theta' = theta + dtheta (normalized)
        3. Updates velocities: vx' = dx/dt, vy' = dy/dt
        
        Args:
            state: Current state [x, y, theta, vx, vy]
            dt: Time step (seconds)
            dx, dy: Linear displacement in x, y
            dtheta: Angular displacement
            
        Returns:
            New state after applying motion model
        """
        # STUDENT IMPLEMENTATION - Task A3
        # ====================================================================
        
        new_state = state.copy()
        
        # Update position
        new_state[0] = state[0] + dx  # x' = x + dx
        new_state[1] = state[1] + dy  # y' = y + dy
        
        # Update orientation and normalize to [-pi, pi]
        new_state[2] = self.normalize_angle(state[2] + dtheta)
        
        # Update velocities (dx/dt, dy/dt)
        if dt > 1e-6:  # Avoid division by zero
            new_state[3] = dx / dt  # vx
            new_state[4] = dy / dt  # vy
        else:
            new_state[3] = 0.0
            new_state[4] = 0.0
        
        return new_state
    
    def measurement_model(self, state: np.ndarray, landmark_id: int) -> np.ndarray:
        """
        Predict measurement given current state and landmark
        
        Task A4: Measurement Model (8 Points)
        
        STUDENT TODO:
        1. Calculate relative position: landmark - robot position
        2. Transform to robot frame using robot orientation
        3. Return relative position in robot frame
        
        Args:
            state: Current state [x, y, theta, vx, vy]
            landmark_id: ID of the landmark to observe
            
        Returns:
            Predicted measurement [rel_x, rel_y] in robot frame
        """
        # STUDENT IMPLEMENTATION - Task A4
        # ====================================================================
        
        # Check if landmark exists
        if landmark_id not in self.landmarks:
            return np.zeros(2)
        
        # Get landmark position in world frame
        lx, ly = self.landmarks[landmark_id]
        
        # Get robot position and orientation
        rx, ry, theta = state[0], state[1], state[2]
        
        # Calculate relative position in world frame
        dx_world = lx - rx
        dy_world = ly - ry
        
        # Transform to robot frame using rotation
        # [rel_x]   [cos(theta)  sin(theta)] [dx_world]
        # [rel_y] = [-sin(theta) cos(theta)] [dy_world]
        cos_theta = np.cos(theta)
        sin_theta = np.sin(theta)
        
        rel_x = cos_theta * dx_world + sin_theta * dy_world
        rel_y = -sin_theta * dx_world + cos_theta * dy_world
        
        return np.array([rel_x, rel_y])
    
    @staticmethod
    def normalize_angle(angle: float) -> float:
        """Normalize angle to [-pi, pi]"""
        while angle > np.pi:
            angle -= 2.0 * np.pi
        while angle < -np.pi:
            angle += 2.0 * np.pi
        return angle
    
    def predict(self, dt: float, dx: float, dy: float, dtheta: float):
        """
        Kalman Filter Prediction Step (Time Update)
        
        Task A5: Predict Step (10 Points)
        
        STUDENT TODO:
        1. Generate sigma points from current state and covariance
        2. Propagate each sigma point through motion model
        3. Calculate mean and covariance of predicted sigma points
        4. Add process noise
        5. Update state and covariance estimates
        
        Args:
            dt: Time step (seconds)
            dx, dy: Linear displacement
            dtheta: Angular displacement
        """
        # STUDENT IMPLEMENTATION - Task A5
        # ====================================================================
        
        # Step 1: Generate sigma points from current state and covariance
        sigma_points = self.generate_sigma_points(self.x, self.P)
        
        # Step 2: Propagate each sigma point through the process model
        sigma_points_pred = []
        for sp in sigma_points:
            sp_pred = self.process_model(sp, dt, dx, dy, dtheta)
            sigma_points_pred.append(sp_pred)
        
        # Step 3: Calculate predicted mean
        x_pred = np.zeros(self.nx)
        for i, sp in enumerate(sigma_points_pred):
            x_pred += self.Wm[i] * sp
        
        # Normalize angle in predicted mean
        x_pred[2] = self.normalize_angle(x_pred[2])
        
        # Step 4: Calculate predicted covariance
        P_pred = np.zeros((self.nx, self.nx))
        for i, sp in enumerate(sigma_points_pred):
            # Calculate deviation
            diff = sp - x_pred
            # Normalize angle difference
            diff[2] = self.normalize_angle(diff[2])
            # Add weighted outer product
            P_pred += self.Wc[i] * np.outer(diff, diff)
        
        # Step 5: Add process noise
        P_pred += self.Q
        
        # Update state and covariance
        self.x = x_pred
        self.P = P_pred
    
    def update(self, landmark_observations: List[Tuple[int, float, float, float]]):
        """
        Kalman Filter Update Step (Measurement Update)
        
        Task A6: Update Step (7 Points)
        
        STUDENT TODO:
        1. Generate sigma points
        2. Transform through measurement model
        3. Calculate predicted measurement mean
        4. Calculate measurement and cross-covariance
        5. Compute Kalman gain
        6. Update state with innovation
        7. Update covariance
        
        Args:
            landmark_observations: List of (id, x, y, noise_cov) tuples
                - id: landmark ID
                - x, y: observed position in robot frame
                - noise_cov: observation noise covariance (not used, we use R)
        """
        # STUDENT IMPLEMENTATION - Task A6
        # ====================================================================
        
        if not landmark_observations:
            return
        
        # Process each landmark observation
        for obs in landmark_observations:
            landmark_id, obs_x, obs_y, _ = obs
            
            # Skip if landmark is not known
            if landmark_id not in self.landmarks:
                continue
            
            # Step 1: Generate sigma points from current state
            sigma_points = self.generate_sigma_points(self.x, self.P)
            
            # Step 2: Transform sigma points through measurement model
            Z_sigma = []
            for sp in sigma_points:
                z_pred = self.measurement_model(sp, landmark_id)
                Z_sigma.append(z_pred)
            
            # Step 3: Calculate predicted measurement mean
            z_mean = np.zeros(self.nz)
            for i, z in enumerate(Z_sigma):
                z_mean += self.Wm[i] * z
            
            # Step 4a: Calculate measurement covariance (innovation covariance)
            P_zz = np.zeros((self.nz, self.nz))
            for i, z in enumerate(Z_sigma):
                diff = z - z_mean
                P_zz += self.Wc[i] * np.outer(diff, diff)
            
            # Add measurement noise
            P_zz += self.R
            
            # Step 4b: Calculate cross-covariance
            P_xz = np.zeros((self.nx, self.nz))
            for i, (sp, z) in enumerate(zip(sigma_points, Z_sigma)):
                x_diff = sp - self.x
                x_diff[2] = self.normalize_angle(x_diff[2])  # Normalize angle
                z_diff = z - z_mean
                P_xz += self.Wc[i] * np.outer(x_diff, z_diff)
            
            # Step 5: Compute Kalman gain
            try:
                K = P_xz @ np.linalg.inv(P_zz)
            except np.linalg.LinAlgError:
                print("Warning: Singular matrix in Kalman gain computation")
                continue
            
            # Step 6: Update state with innovation
            z_obs = np.array([obs_x, obs_y])
            innovation = z_obs - z_mean
            self.x = self.x + K @ innovation
            
            # Normalize angle after update
            self.x[2] = self.normalize_angle(self.x[2])
            
            # Step 7: Update covariance
            self.P = self.P - K @ P_zz @ K.T
            
            # Ensure P remains symmetric and positive definite
            self.P = (self.P + self.P.T) / 2.0
    
    def set_landmarks(self, landmarks: Dict[int, Tuple[float, float]]):
        """Set landmark positions"""
        self.landmarks = landmarks
    
    def has_landmark(self, landmark_id: int) -> bool:
        """Check if landmark exists"""
        return landmark_id in self.landmarks
    
    def get_state(self) -> np.ndarray:
        """Get estimated state [x, y, theta, vx, vy]"""
        return self.x.copy()
    
    def get_covariance(self) -> np.ndarray:
        """Get state covariance matrix"""
        return self.P.copy()
    
    def get_position(self) -> Tuple[float, float]:
        """Get position estimate (x, y)"""
        return (self.x[0], self.x[1])
    
    def get_orientation(self) -> float:
        """Get orientation estimate (theta)"""
        return self.x[2]
    
    def get_velocity(self) -> Tuple[float, float]:
        """Get velocity estimate (vx, vy)"""
        return (self.x[3], self.x[4])

